File: day19_part1_failed_3/main.py
import itertools
from collections import defaultdict

# Step 1: Parse input
def parse_input(filename):
    with open(filename) as file:
        raw_data = file.read().strip().split('\n\n')
    scanners = []
    for data in raw_data:
        beacons = [tuple(map(int, line.split(','))) for line in data.split('\n')[1:]]
        scanners.append(beacons)
    return scanners

# Step 2: Generate distance vectors
def generate_distance_vectors(beacons):
    return {(x2 - x1, y2 - y1, z2 - z1)
            for x1, y1, z1 in beacons
            for x2, y2, z2 in beacons}

# Step 3: Find overlapping scanners
def find_overlaps(scanners):
    overlap_info = {}
    for i, scanner_a in enumerate(scanners):
        for j, scanner_b in enumerate(scanners):
            if i >= j:
                continue
            vectors_a = generate_distance_vectors(scanner_a)
            vectors_b = generate_distance_vectors(scanner_b)
            common = vectors_a & vectors_b
            if len(common) >= 66:  # Choose 2 from 12 common beacons
                overlap_info[(i, j)] = common
    return overlap_info

# Step 4: Determine orientation and position
def determine_orientation_and_position(scanner_a, scanner_b, common_vectors):
    for perm in itertools.permutations((0, 1, 2)):
        for flips in itertools.product((-1, 1), repeat=3):
            transformed_b = {tuple(coord[perm[i]] * flips[i] for i in range(3)) for coord in scanner_b}
            vectors_b = generate_distance_vectors(transformed_b)
            if vectors_b & common_vectors == common_vectors:
                for beacon_a in scanner_a:
                    for beacon_b in transformed_b:
                        translation = tuple(beacon_a[i] - beacon_b[i] for i in range(3))
                        translated_b = {tuple(beacon[i] + translation[i] for i in range(3)) for beacon in transformed_b}
                        if len(translated_b & set(scanner_a)) >= 12:
                            return translated_b, translation
    return set(), (0, 0, 0)

# Step 5: Transform beacon coordinates
def transform_coordinates(scanners, overlaps):
    aligned = {0: scanners[0]}
    translations = {0: (0, 0, 0)}
    while len(aligned) < len(scanners):
        for (i, j), common_vectors in overlaps.items():
            if i in aligned and j not in aligned:
                transformed_b, translation = determine_orientation_and_position(aligned[i], scanners[j], common_vectors)
                if transformed_b:
                    aligned[j] = transformed_b
                    translations[j] = translation
    return aligned, translations

# Step 6: Combine all beacons
def combine_beacons(aligned):
    global_beacons = set()
    for beacons in aligned.values():
        global_beacons.update(beacons)
    return global_beacons

# Main execution
if __name__ == "__main__":
    scanners = parse_input("day19_part1/input.txt")
    overlaps = find_overlaps(scanners)
    aligned_scanners, scanner_positions = transform_coordinates(scanners, overlaps)
    global_beacons = combine_beacons(aligned_scanners)
    print(f"Number of beacons: {len(global_beacons)}")

File: day19_part1_failed_3/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
)

// Beacon represents a 3D point.
type Beacon struct{ x, y, z int }

// DistanceVector represents a vector between two Beacons.
type DistanceVector struct{ x, y, z int }

// ScannerData holds the beacons seen by a scanner.
type ScannerData struct {
	beacons []Beacon
}

// NewBeacon creates a new Beacon from coordinates.
func NewBeacon(coords []int) Beacon {
	return Beacon{coords[0], coords[1], coords[2]}
}

// Subtract creates a DistanceVector representing the distance from b to a.
func (a Beacon) Subtract(b Beacon) DistanceVector {
	return DistanceVector{a.x - b.x, a.y - b.y, a.z - b.z}
}

// Add applies a DistanceVector to a Beacon.
func (a Beacon) Add(v DistanceVector) Beacon {
	return Beacon{a.x + v.x, a.y + v.y, a.z + v.z}
}

// DistanceVectors returns all DistanceVectors between beacons in the ScannerData.
func (sd *ScannerData) DistanceVectors() map[DistanceVector]bool {
	vectors := make(map[DistanceVector]bool)
	for _, b1 := range sd.beacons {
		for _, b2 := range sd.beacons {
			if b1 != b2 {
				vectors[b1.Subtract(b2)] = true
			}
		}
	}
	return vectors
}

// ReadInput reads scanner data from a file.
func ReadInput(filename string) ([]ScannerData, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scannerData := []ScannerData{}
	currentScannerData := ScannerData{}

	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			scannerData = append(scannerData, currentScannerData)
			currentScannerData = ScannerData{}
			continue
		}

		if strings.HasPrefix(line, "---") {
			continue
		}

		coordsStr := strings.Split(line, ",")
		coords := make([]int, 3)
		for i, coordStr := range coordsStr {
			coord, err := strconv.Atoi(coordStr)
			if err != nil {
				return nil, err
			}
			coords[i] = coord
		}

		currentScannerData.beacons = append(currentScannerData.beacons, NewBeacon(coords))
	}

	if len(currentScannerData.beacons) > 0 {
		scannerData = append(scannerData, currentScannerData)
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return scannerData, nil
}

// FindOverlaps identifies scanners with overlapping detection vectors.
func FindOverlaps(scanners []ScannerData) map[int]map[int][]DistanceVector {
	overlaps := make(map[int]map[int][]DistanceVector)

	// For each scanner, precompute the set of distance vectors between all pairs of beacons.
	precomputedVectors := make([]map[DistanceVector]bool, len(scanners))
	for i, scanner := range scanners {
		precomputedVectors[i] = scanner.DistanceVectors()
	}

	// Compare the distance vectors between each pair of scanners to find overlaps.
	for i := 0; i < len(scanners); i++ {
		for j := i + 1; j < len(scanners); j++ {
			// Initialize the nested map for the current scanner pair if it doesn't exist.
			if overlaps[i] == nil {
				overlaps[i] = make(map[int][]DistanceVector)
			}

			// Find common distance vectors between scanner i and scanner j.
			for vec := range precomputedVectors[i] {
				if precomputedVectors[j][vec] {
					overlaps[i][j] = append(overlaps[i][j], vec)
				}
			}
		}
	}

	return overlaps
}

// DetermineOrientationAndPosition determines how to align one scanner to another.
func DetermineOrientationAndPosition(scannerA, scannerB ScannerData, overlaps []DistanceVector) ([][]int, Beacon) {
	// Code to determine orientation (as a rotation matrix) and position goes here...
	return [][]int{{1, 0, 0}, {0, 1, 0}, {0, 0, 1}}, Beacon{0, 0, 0} // Example output
}

// TransformCoordinates applies a rotation and translation to align scanner B to A.
func TransformCoordinates(beacons []Beacon, rotation [][]int, translation Beacon) []Beacon {
	transformed := make([]Beacon, len(beacons))
	// Code to transform coordinates goes here...
	return transformed
}

// CombineBeacons merges all beacons into a global set, removing duplicates.
func CombineBeacons(scanners []ScannerData, alignments map[int]map[int][][]int, translations map[int]map[int]Beacon) []Beacon {
	globalBeacons := make([]Beacon, 0)
	// Code to combine beacons goes here...
	return globalBeacons
}

func main() {
	// Read scanner data from the input file.
	scannerData, err := ReadInput("input.txt")
	if err != nil {
		panic(err)
	}

	// Find overlapping scanners.
	overlaps := FindOverlaps(scannerData)

	// Determine orientations and positions for all scanners relative to scanner 0.
	alignments := make(map[int]map[int][][]int)  // Maps scanner pairs to rotation matrices.
	translations := make(map[int]map[int]Beacon) // Maps scanner pairs to translation vectors.

	var wg sync.WaitGroup
	for i, scannerA := range scannerData {
		for j, scannerB := range scannerData {
			if i == j || len(overlaps[i][j]) == 0 {
				continue
			}
			wg.Add(1)
			go func(i, j int, scannerA, scannerB ScannerData) {
				defer wg.Done()
				rotation, translation := DetermineOrientationAndPosition(scannerA, scannerB, overlaps[i][j])
				alignments[i][j] = rotation
				translations[i][j] = translation
			}(i, j, scannerA, scannerB)
		}
	}
	wg.Wait()

	// Apply transformations and combine all beacons.
	globalBeacons := CombineBeacons(scannerData, alignments, translations)
	fmt.Printf("Number of beacons: %d\n", len(globalBeacons))
}
